[WIP] Fstrings

brianbok opened the pull request.

brianbok committed:
Added test file

brianbok committed:
FString Expression

brianbok committed:
First test passing

brianbok committed:
WIP Parser

brianbok committed:
More tests

brianbok committed:
Escaping braces

brianbok committed:
Merge branch 'master' into fstrings

jakebailey commented on src/Parsing/Impl/FStringParser.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        private void AddStringToChildren(string x) {
+            var parser = Parser.CreateParser(new MemoryStream(Encoding.UTF8.GetBytes(x)), PythonLanguageVersion.V36);
```
The language version is definitely something you'll want to accept in the constructor from the parent parser. I think you may also just want to use a `StringReader` here (since the text has already been converted into a C# string) and don't need to make an assumption about the encoding.

Commit:
De-duplicate completions  (#602)

* Part 7

* Buildable

* PR feedback

* Merge conflict

* Fix #446

* Fix #446

* Part 8

* Part 9

* Buildable

* Part 10

* Part 11

* Part 12

* Buildable

* Part 14

* First passing test

* Simplify configuration

* Style

* Fix test and move code to folders

* Builtins import

* Fix #470

* Fluents

* Add search path

* Import analysis, part I

* Simplify builtins handling

* Remove IMember

* Handle import specific

* More tests

* Add typeshed

* Renames

* Make sure lazy modules are loaded

* Renames

* Move/rename

* Rework importing

* Derivation rework

* Part 2

* Part 3

* Buildable

* Module members

* Async walk

* Imports test pass

* Remove lazy types

* Fix from import

* Stubs

* Double overloads

* Fix datetime test

* Couple more tests + fluents

* Few more tests

* Additionl test + union type

* Built-in scrape tests

* Full stdlib scrape test

* Complete async AST walker

* Conditional defines test + variable loc cleanup

* More stub tests
Fix stub loading for packages (port from DDG)
Split walker into multiple files

* Add some (broken mostly) tests from DDG

* Move document tests

* Function arg eval, part I

* Instance/factory

* Builds

* Test fixes

* Fix static and instance call eval

* More tests

* More ported tests

* Specialize builtin functions

* Make walkers common and handle nested functions

* Moar tests

* Parser fixes + more tests

* Handle negative numbers

* Fix null ref

* Basic list support

* Few more list tests

* Basic iterators

* Support __iter__

* Iterators

* Fix couple of tests

* Add decorator test

* Generics, part I

* Generics, part 2

* Generics, part 3

* Basic TypeVar test

* Typings, part 4

* Fix test

* Generics, part 6

* Generics, part 7

* More tests (failing)

* Forward ref fixes

* Reorg

* Improve symbol resolution + test fixes

* Test fixes

* Dictionary, part I

* Part 11

* Fix test

* Tests

* Tests

* More dict work

* List ctor

* Skip some tests for now

* Fix iterators

* Tuple slicing

* Polish type comparo in return types

* Add Mapping and tests

* Add Iterable

* Fix typo

* Add Iterator[T] + test

* Simplify typing types

* Class reduction

* Fix tests

* Test fix

* Handle 'with' statement

* Handle try-except

* Class method inheritance + NewType

* Container types

* Containers test

* Tests

* Handle generic type alias

* Named tuple

* Global/non-local

* Handle tuples in for
Handle custom iterators

* Basic generator

* Any/AnyStr

* Test fixes

* Type/Optional/etc handling

* Proper doc population

* Tests + range

* Argument match

* Basic argset and diagnostics

* Argset tests

* Exclude WIP

* Exclude WIP

* Arg eval

* Arg match, part 2

* Tests and generic arg comparisons

* Function eval with arguments

* Baselines

* Fix test

* Undo AST formatting change and update baseline

* LS cleanup 1

* Fix list ctor argument unpacking

* Cleanup 2

* Builds

* Partial completions

* Partial

* Partial

* Simple test

* Tests

* Basic startup

* Clean up a bit

* Remove debug code

* Port formatter tests

* Fix tokenizer crash

* Async fixes

* Hover

* Basic hover

* Adjust expression options

* Basic signature help

* Fix class/instance

* Update test

* Fix builtin creation exception

* Fix tests

* Actually provide declared module

* Completion test (partial)

* Undo

* Fix null await
Fix override completions + test

* Exports filtering
Prevent augmenting imported types

* Filter variables & exports

* Ported tests

* Test fixes

* More ported tests

* Fix exception completions

* Import completions

* Scope completions

* With completions

* Test fixes

* WIP

* Test fix

* Better arg match

* Temp disable WIP

* First cut

* Fix type leak

* WIP

* Remove ConfigureAwait and handle canceled and failed in the analysis notifications

* WIP

* Generic class base

* Generic forward reference resolution

* Suppress completion in strings + test

* Prevent recursion on generic resolution
Better match arguments

* Handle call expression in generics

* Relax condition as it happens in tensorflow

* Fix typeshed version search
Make writing cached modules async
Fix module doc fetching

* Hover tests

* Fix prom import hover

* Hover tests

* Synchronize test cache writing

* First cut

* Test

* Fixes

* Add tests for os.path
Null ref fix

* Fix cache check

* Improve resolution of builtins and typing in stubs

* Merge tests

* Add ntst for requests

* Handle typeshed better

* Fix custom stub handling

* Better sync

* Move files

* Fix parameter locations

* Hover improvement

* PEP hints

* One more test for PEP hints

* Better handle hover over import as

* Text based generic constraints

* Handle with better with generic stubs

* Undo debug

* Handle non-binary open()
Temporary fix 'with' handler since we haven't specialized IO/TextIO/BinaryIO yet.

* Output syntax errors

* Properly clear

* - Fix async issue with analysis completion
- Clean up diagnostics service interface
- Use real DS in tests

* Use proper scope when analyzing module

* Severity mapping and reporting

* Add publishing test
Add NSubstitute
Move all services to the same namespace.

* Unused var

* Test forced publish on close

* Fix typo

* Update test framework

* Import location

* Remove incorrect reference

* Diagnostic severity mapping test
Fix post-mortem earlier PR comment

* Minor fixes

* Move interface to the main class part

* Flicker reduction

* De-duplicate completions

* Fix os.path in typeshed

jakebailey committed:
Fix empty signature tooltip after function call (#616)

Fixes #614.

Pulled this from the old LS; it wasn't carried over.

Commit:
Port import tests from old LS (#606)

* Part 8

* Part 9

* Buildable

* Part 10

* Part 11

* Part 12

* Buildable

* Part 14

* First passing test

* Simplify configuration

* Style

* Fix test and move code to folders

* Builtins import

* Fix #470

* Fluents

* Add search path

* Import analysis, part I

* Simplify builtins handling

* Remove IMember

* Handle import specific

* More tests

* Add typeshed

* Renames

* Make sure lazy modules are loaded

* Renames

* Move/rename

* Rework importing

* Derivation rework

* Part 2

* Part 3

* Buildable

* Module members

* Async walk

* Imports test pass

* Remove lazy types

* Fix from import

* Stubs

* Double overloads

* Fix datetime test

* Couple more tests + fluents

* Few more tests

* Additionl test + union type

* Built-in scrape tests

* Full stdlib scrape test

* Complete async AST walker

* Conditional defines test + variable loc cleanup

* More stub tests
Fix stub loading for packages (port from DDG)
Split walker into multiple files

* Add some (broken mostly) tests from DDG

* Move document tests

* Function arg eval, part I

* Instance/factory

* Builds

* Test fixes

* Fix static and instance call eval

* More tests

* More ported tests

* Specialize builtin functions

* Make walkers common and handle nested functions

* Moar tests

* Parser fixes + more tests

* Handle negative numbers

* Fix null ref

* Basic list support

* Few more list tests

* Basic iterators

* Support __iter__

* Iterators

* Fix couple of tests

* Add decorator test

* Generics, part I

* Generics, part 2

* Generics, part 3

* Basic TypeVar test

* Typings, part 4

* Fix test

* Generics, part 6

* Generics, part 7

* More tests (failing)

* Forward ref fixes

* Reorg

* Improve symbol resolution + test fixes

* Test fixes

* Dictionary, part I

* Part 11

* Fix test

* Tests

* Tests

* More dict work

* List ctor

* Skip some tests for now

* Fix iterators

* Tuple slicing

* Polish type comparo in return types

* Add Mapping and tests

* Add Iterable

* Fix typo

* Add Iterator[T] + test

* Simplify typing types

* Class reduction

* Fix tests

* Test fix

* Handle 'with' statement

* Handle try-except

* Class method inheritance + NewType

* Container types

* Containers test

* Tests

* Handle generic type alias

* Named tuple

* Global/non-local

* Handle tuples in for
Handle custom iterators

* Basic generator

* Any/AnyStr

* Test fixes

* Type/Optional/etc handling

* Proper doc population

* Tests + range

* Argument match

* Basic argset and diagnostics

* Argset tests

* Exclude WIP

* Exclude WIP

* Arg eval

* Arg match, part 2

* Tests and generic arg comparisons

* Function eval with arguments

* Baselines

* Fix test

* Undo AST formatting change and update baseline

* LS cleanup 1

* Fix list ctor argument unpacking

* Cleanup 2

* Builds

* Partial completions

* Partial

* Partial

* Simple test

* Tests

* Basic startup

* Clean up a bit

* Remove debug code

* Port formatter tests

* Fix tokenizer crash

* Async fixes

* Hover

* Basic hover

* Adjust expression options

* Basic signature help

* Fix class/instance

* Update test

* Fix builtin creation exception

* Fix tests

* Actually provide declared module

* Completion test (partial)

* Undo

* Fix null await
Fix override completions + test

* Exports filtering
Prevent augmenting imported types

* Filter variables & exports

* Ported tests

* Test fixes

* More ported tests

* Fix exception completions

* Import completions

* Scope completions

* With completions

* Test fixes

* WIP

* Test fix

* Better arg match

* Temp disable WIP

* First cut

* Fix type leak

* WIP

* Remove ConfigureAwait and handle canceled and failed in the analysis notifications

* WIP

* Generic class base

* Generic forward reference resolution

* Suppress completion in strings + test

* Prevent recursion on generic resolution
Better match arguments

* Handle call expression in generics

* Relax condition as it happens in tensorflow

* Fix typeshed version search
Make writing cached modules async
Fix module doc fetching

* Hover tests

* Fix prom import hover

* Hover tests

* Synchronize test cache writing

* First cut

* Test

* Fixes

* Add tests for os.path
Null ref fix

* Fix cache check

* Improve resolution of builtins and typing in stubs

* Merge tests

* Add ntst for requests

* Handle typeshed better

* Fix custom stub handling

* Better sync

* Move files

* Fix parameter locations

* Hover improvement

* PEP hints

* One more test for PEP hints

* Better handle hover over import as

* Text based generic constraints

* Handle with better with generic stubs

* Undo debug

* Handle non-binary open()
Temporary fix 'with' handler since we haven't specialized IO/TextIO/BinaryIO yet.

* Output syntax errors

* Properly clear

* - Fix async issue with analysis completion
- Clean up diagnostics service interface
- Use real DS in tests

* Use proper scope when analyzing module

* Severity mapping and reporting

* Add publishing test
Add NSubstitute
Move all services to the same namespace.

* Unused var

* Test forced publish on close

* Fix typo

* Update test framework

* Import location

* Remove incorrect reference

* Diagnostic severity mapping test
Fix post-mortem earlier PR comment

* Minor fixes

* Move interface to the main class part

* Flicker reduction

* - Correct reported unresolved import name
- Add tests

* PR feedback

* Port 2 tests

* More tests

* More tests

* Fix typing and package import test

* Last test

* Call base instead

Commit:
Resolve issue when specific types shared data after creation off generic template (#600)

* Buildable

* Part 14

* First passing test

* Simplify configuration

* Style

* Fix test and move code to folders

* Builtins import

* Fix #470

* Fluents

* Add search path

* Import analysis, part I

* Simplify builtins handling

* Remove IMember

* Handle import specific

* More tests

* Add typeshed

* Renames

* Make sure lazy modules are loaded

* Renames

* Move/rename

* Rework importing

* Derivation rework

* Part 2

* Part 3

* Buildable

* Module members

* Async walk

* Imports test pass

* Remove lazy types

* Fix from import

* Stubs

* Double overloads

* Fix datetime test

* Couple more tests + fluents

* Few more tests

* Additionl test + union type

* Built-in scrape tests

* Full stdlib scrape test

* Complete async AST walker

* Conditional defines test + variable loc cleanup

* More stub tests
Fix stub loading for packages (port from DDG)
Split walker into multiple files

* Add some (broken mostly) tests from DDG

* Move document tests

* Function arg eval, part I

* Instance/factory

* Builds

* Test fixes

* Fix static and instance call eval

* More tests

* More ported tests

* Specialize builtin functions

* Make walkers common and handle nested functions

* Moar tests

* Parser fixes + more tests

* Handle negative numbers

* Fix null ref

* Basic list support

* Few more list tests

* Basic iterators

* Support __iter__

* Iterators

* Fix couple of tests

* Add decorator test

* Generics, part I

* Generics, part 2

* Generics, part 3

* Basic TypeVar test

* Typings, part 4

* Fix test

* Generics, part 6

* Generics, part 7

* More tests (failing)

* Forward ref fixes

* Reorg

* Improve symbol resolution + test fixes

* Test fixes

* Dictionary, part I

* Part 11

* Fix test

* Tests

* Tests

* More dict work

* List ctor

* Skip some tests for now

* Fix iterators

* Tuple slicing

* Polish type comparo in return types

* Add Mapping and tests

* Add Iterable

* Fix typo

* Add Iterator[T] + test

* Simplify typing types

* Class reduction

* Fix tests

* Test fix

* Handle 'with' statement

* Handle try-except

* Class method inheritance + NewType

* Container types

* Containers test

* Tests

* Handle generic type alias

* Named tuple

* Global/non-local

* Handle tuples in for
Handle custom iterators

* Basic generator

* Any/AnyStr

* Test fixes

* Type/Optional/etc handling

* Proper doc population

* Tests + range

* Argument match

* Basic argset and diagnostics

* Argset tests

* Exclude WIP

* Exclude WIP

* Arg eval

* Arg match, part 2

* Tests and generic arg comparisons

* Function eval with arguments

* Baselines

* Fix test

* Undo AST formatting change and update baseline

* LS cleanup 1

* Fix list ctor argument unpacking

* Cleanup 2

* Builds

* Partial completions

* Partial

* Partial

* Simple test

* Tests

* Basic startup

* Clean up a bit

* Remove debug code

* Port formatter tests

* Fix tokenizer crash

* Async fixes

* Hover

* Basic hover

* Adjust expression options

* Basic signature help

* Fix class/instance

* Update test

* Fix builtin creation exception

* Fix tests

* Actually provide declared module

* Completion test (partial)

* Undo

* Fix null await
Fix override completions + test

* Exports filtering
Prevent augmenting imported types

* Filter variables & exports

* Ported tests

* Test fixes

* More ported tests

* Fix exception completions

* Import completions

* Scope completions

* With completions

* Test fixes

* WIP

* Test fix

* Better arg match

* Temp disable WIP

* First cut

* Fix type leak

* WIP

* Remove ConfigureAwait and handle canceled and failed in the analysis notifications

* WIP

* Generic class base

* Generic forward reference resolution

* Suppress completion in strings + test

* Prevent recursion on generic resolution
Better match arguments

* Handle call expression in generics

* Relax condition as it happens in tensorflow

* Fix typeshed version search
Make writing cached modules async
Fix module doc fetching

* Hover tests

* Fix prom import hover

* Hover tests

* Synchronize test cache writing

* First cut

* Test

* Fixes

* Add tests for os.path
Null ref fix

* Fix cache check

* Improve resolution of builtins and typing in stubs

* Merge tests

* Add ntst for requests

* Handle typeshed better

* Fix custom stub handling

* Better sync

* Move files

* Fix parameter locations

* Hover improvement

* PEP hints

* One more test for PEP hints

* Better handle hover over import as

* Text based generic constraints

* Handle with better with generic stubs

* Undo debug

* Handle non-binary open()
Temporary fix 'with' handler since we haven't specialized IO/TextIO/BinaryIO yet.

* Output syntax errors

* Properly clear

* - Fix async issue with analysis completion
- Clean up diagnostics service interface
- Use real DS in tests

* Use proper scope when analyzing module

* Severity mapping and reporting

* Add publishing test
Add NSubstitute
Move all services to the same namespace.

* Unused var

* Test forced publish on close

* Fix typo

* Update test framework

* Import location

* Remove incorrect reference

* Diagnostic severity mapping test
Fix post-mortem earlier PR comment

* Minor fixes

* Better handle return types in classes created from templates.

* Move interface to the main class part

* Dynamic return type

* Add hover and signature tests

* Baseline update

* Flicker reduction

* - Correct reported unresolved import name
- Add tests

* PR feedback

* Resolve merge issues

* Fix completion doc + test

* Restore formatting

* Bunch of null checks

* Fix generic base classes
Add tests and changes to couple more cases

* PR feedback

AlexanderSher committed:
Add dependency resolution to the analyzer (#621)

* Add dependency resolution to the analyzer

* Don't wait on missing Typeshed files

* Merge "De-duplicate completions (#602)"

* Fix empty signature tooltip after function call (#616)

Fixes #614.

* - Address CR Comments
- Merge "Port import tests from old LS (#606)"

* Resolve issue when specific types shared data after creation off generic template (#600)

* Buildable

* Part 14

* First passing test

* Simplify configuration

* Style

* Fix test and move code to folders

* Builtins import

* Fix #470

* Fluents

* Add search path

* Import analysis, part I

* Simplify builtins handling

* Remove IMember

* Handle import specific

* More tests

* Add typeshed

* Renames

* Make sure lazy modules are loaded

* Renames

* Move/rename

* Rework importing

* Derivation rework

* Part 2

* Part 3

* Buildable

* Module members

* Async walk

* Imports test pass

* Remove lazy types

* Fix from import

* Stubs

* Double overloads

* Fix datetime test

* Couple more tests + fluents

* Few more tests

* Additionl test + union type

* Built-in scrape tests

* Full stdlib scrape test

* Complete async AST walker

* Conditional defines test + variable loc cleanup

* More stub tests
Fix stub loading for packages (port from DDG)
Split walker into multiple files

* Add some (broken mostly) tests from DDG

* Move document tests

* Function arg eval, part I

* Instance/factory

* Builds

* Test fixes

* Fix static and instance call eval

* More tests

* More ported tests

* Specialize builtin functions

* Make walkers common and handle nested functions

* Moar tests

* Parser fixes + more tests

* Handle negative numbers

* Fix null ref

* Basic list support

* Few more list tests

* Basic iterators

* Support __iter__

* Iterators

* Fix couple of tests

* Add decorator test

* Generics, part I

* Generics, part 2

* Generics, part 3

* Basic TypeVar test

* Typings, part 4

* Fix test

* Generics, part 6

* Generics, part 7

* More tests (failing)

* Forward ref fixes

* Reorg

* Improve symbol resolution + test fixes

* Test fixes

* Dictionary, part I

* Part 11

* Fix test

* Tests

* Tests

* More dict work

* List ctor

* Skip some tests for now

* Fix iterators

* Tuple slicing

* Polish type comparo in return types

* Add Mapping and tests

* Add Iterable

* Fix typo

* Add Iterator[T] + test

* Simplify typing types

* Class reduction

* Fix tests

* Test fix

* Handle 'with' statement

* Handle try-except

* Class method inheritance + NewType

* Container types

* Containers test

* Tests

* Handle generic type alias

* Named tuple

* Global/non-local

* Handle tuples in for
Handle custom iterators

* Basic generator

* Any/AnyStr

* Test fixes

* Type/Optional/etc handling

* Proper doc population

* Tests + range

* Argument match

* Basic argset and diagnostics

* Argset tests

* Exclude WIP

* Exclude WIP

* Arg eval

* Arg match, part 2

* Tests and generic arg comparisons

* Function eval with arguments

* Baselines

* Fix test

* Undo AST formatting change and update baseline

* LS cleanup 1

* Fix list ctor argument unpacking

* Cleanup 2

* Builds

* Partial completions

* Partial

* Partial

* Simple test

* Tests

* Basic startup

* Clean up a bit

* Remove debug code

* Port formatter tests

* Fix tokenizer crash

* Async fixes

* Hover

* Basic hover

* Adjust expression options

* Basic signature help

* Fix class/instance

* Update test

* Fix builtin creation exception

* Fix tests

* Actually provide declared module

* Completion test (partial)

* Undo

* Fix null await
Fix override completions + test

* Exports filtering
Prevent augmenting imported types

* Filter variables & exports

* Ported tests

* Test fixes

* More ported tests

* Fix exception completions

* Import completions

* Scope completions

* With completions

* Test fixes

* WIP

* Test fix

* Better arg match

* Temp disable WIP

* First cut

* Fix type leak

* WIP

* Remove ConfigureAwait and handle canceled and failed in the analysis notifications

* WIP

* Generic class base

* Generic forward reference resolution

* Suppress completion in strings + test

* Prevent recursion on generic resolution
Better match arguments

* Handle call expression in generics

* Relax condition as it happens in tensorflow

* Fix typeshed version search
Make writing cached modules async
Fix module doc fetching

* Hover tests

* Fix prom import hover

* Hover tests

* Synchronize test cache writing

* First cut

* Test

* Fixes

* Add tests for os.path
Null ref fix

* Fix cache check

* Improve resolution of builtins and typing in stubs

* Merge tests

* Add ntst for requests

* Handle typeshed better

* Fix custom stub handling

* Better sync

* Move files

* Fix parameter locations

* Hover improvement

* PEP hints

* One more test for PEP hints

* Better handle hover over import as

* Text based generic constraints

* Handle with better with generic stubs

* Undo debug

* Handle non-binary open()
Temporary fix 'with' handler since we haven't specialized IO/TextIO/BinaryIO yet.

* Output syntax errors

* Properly clear

* - Fix async issue with analysis completion
- Clean up diagnostics service interface
- Use real DS in tests

* Use proper scope when analyzing module

* Severity mapping and reporting

* Add publishing test
Add NSubstitute
Move all services to the same namespace.

* Unused var

* Test forced publish on close

* Fix typo

* Update test framework

* Import location

* Remove incorrect reference

* Diagnostic severity mapping test
Fix post-mortem earlier PR comment

* Minor fixes

* Better handle return types in classes created from templates.

* Move interface to the main class part

* Dynamic return type

* Add hover and signature tests

* Baseline update

* Flicker reduction

* - Correct reported unresolved import name
- Add tests

* PR feedback

* Resolve merge issues

* Fix completion doc + test

* Restore formatting

* Bunch of null checks

* Fix generic base classes
Add tests and changes to couple more cases

* PR feedback

* Remove double assigning

* Don't allow non-rooted modules to be added as part of reload

AlexanderSher committed:
Build fix (#630)

Commit:
Diagnostics should be reported only for changed files (#624)

* Diagnostics reporting

* Handle standole files better

* Merge issues

* Fix test

* Add await doc.GetAstAsync();

* Adjust time

Commit:
Fix #601 (#622)

* Fix #601

* Merge issues

* Merge branch 'master' of https://github.com/Microsoft/python-language-server into 601

* Using

AlexanderSher committed:
Distinguish library module from user module based on path resolver roots (#631)

Commit:
Fix hovering over parts of 'from A.B import C` (#623)

* Fix #601

* Fix #607

* Update test

* Fix version conditionals
Implement platform handling
Specialize os.path

* Null checks

* Merge issues

* Merge branch 'master' of https://github.com/Microsoft/python-language-server into 601

* Using

* Merge master

AlexanderSher committed:
Return stub as module if it is requested as module (#632)

jakebailey committed:
fix null check in switch case (#634)

AlexanderSher committed:
Fix #421: Completions not available for imported functions (#641)

brianbok committed:
Merge branch 'master' into fstrings

Commit:
Fix processing of inner functions + remove async from analysis (#649)

* Optimistically return instance if callable turns out to be a property or a dynamic member.

* Remove async
Better handle inner functions

Commit:
Don't  squiggle new lines after incomplete member expression (#648)

* Sqiggle prev token

* Don't squiggle new lines

* Adjust per feedback

* Formatting

jakebailey committed:
Fix signature trigger characters (#654)

Fixes #651.

Tested in VS Code, works the same.

jakebailey committed:
Remove path cache file (#653)

Closes #652.

brianbok committed:
implement abstract node

jakebailey committed:
Port test from #312 to new analysis (#657)

Grabbing this test for completeness' sake.

Commit:
Upgrade to 2.2, fix projects (#656)

brianbok committed:
Use stringreader instead of memorybuffer

brianbok committed:
Case pattern matching

Commit:
Fix goto definition in stubbed modules (#659)

* Merge part I

* Merge part II

* Fix nav to builtins

brianbok committed:
Async Indexer for new LanguageServer (#558)

Adding back SymbolIndex to new LanguageServer

Commit:
Fix NRE (#660)

Commit:
Don't evaluate function body in stubs (#661)

* Fix NRE

* Don't eval stub functions

AlexanderSher committed:
Several bug fixes (#655)

* Fix goto definition in stubbed modules

* PR fixes

* Make stub -> main relationship explicit

* Handle variables separately

* - Prevent several analysis of the same file to run concurrently
- Create initial analysis for every module type (not just the User)
- Add ability so schedule analysis-only dependencies

* Build fix

* - Add logging for the whole PythonAnalyzer session
- Fix cancellation in AsyncAutoResetEvent
- Look for non-analyzed entries after waiting for the _analysisRunningEvent

* Fix linebreaks

Commit:
Repair tests post massive merge (#662)

* Fix NRE

* Don't eval stub functions

* Better stub eval fix

* Merge issues

brianbok committed:
Merge branch 'master' into fstrings

jakebailey committed:
Handle more heading cases in docstring converter (#664)

Fixes #629.

Not the cleanest method; in the future we may need to handle ReST headings better based on the spec, but this covers everything I've seen so far.

AlexanderSher committed:
- Fix the issue with DependencyResolver removing vertices from changed list without them being processed (#665)

- Reduce lock region
- Remove unused awaits

brianbok committed:
Many many tests

Commit:
Fix exception in type annotation parser (#672)

* Fix NRE

* Don't eval stub functions

* Better stub eval fix

* Merge issues

* Fix exception

brianbok committed:
Fixed and refactoring

brianbok committed:
Parsing of concat fstrings

brianbok committed:
Refactoring

brianbok committed:
Adding parsing of conversion and format parameters

Commit:
Reload modules on pip install (#667)

* Fix NRE

* Don't eval stub functions

* Better stub eval fix

* Merge issues

* Reload modules

* Undo a change

* Clear module table on reload

Commit:
Improve handling of Union index and call (#676)

* Fix NRE

* Don't eval stub functions

* Better stub eval fix

* Merge issues

* Fix exception

* Improve union handling

* Move off the lock

Commit:
Fixes occasional 'collection modified' exception (#677)

* Fix NRE

* Don't eval stub functions

* Better stub eval fix

* Merge issues

* Fix exception

* Dont' eval stubs

* Use has flag

Commit:
Fix handling of generics and class constructors that define variables (#666)

* Fix #601

* Fix #607

* Update test

* Fix version conditionals
Implement platform handling
Specialize os.path

* Optional/None

* Make sure constructors are processed

* Null checks

* Merge issues

* Merge branch 'master' of https://github.com/Microsoft/python-language-server into 601

* Using

* Merge master

* Attach object as base to 3x classes

* using

* Partial

* Handle or expression in parameters

* Optimistically return instance if callable turns out to be a property or a dynamic member.

* Remove async
Better handle inner functions

* Pass ctor arguments

* Test updates

* Tests

* Fix SO

* Generic functions - partial

* Generic args

* Tests and formatting

* Fix NRE

* Don't eval stub functions

* Fix doc strings

* Handle standalone generic + enable test
Split typing ang generics tests

* Tests update

* More gotodef tests

* Property handle constructors defining members

* fix function doc

* Debug code

* Add as name check

* Fix import

* Add AsyncLocal

* Revert "Add AsyncLocal"

This reverts commit 5a808b62cd48820668fd05e96d3e496b5d636e2f.

Commit:
Сapture Python output reliably (#673)

* Fix NRE

* Don't eval stub functions

* Better stub eval fix

* Merge issues

* Implement queue

* Make operation atomic

* Port services from RTVS

* Add service to one more place

* Proper path compare

* PR feedback

Commit:
Treat default value None as nothing (#680)

* Fix #604

* PR feedback

brianbok committed:
Refactoring parsing

brianbok committed:
Error handling

brianbok committed:
Nit

brianbok committed:
Newline nit

Commit:
Fix #668 (partial) (#683)

Commit:
Wire up restart of analysis on changes in configuration (#684)

* Wire up config changes

* Fix error

* Log null settings

brianbok committed:
Concat of strings refactor

brianbok committed:
PR comments fixes

brianbok committed:
Error handling

brianbok committed:
Test more fstring errors

brianbok committed:
Parsing logic

Commit:
Fix Requests test (#694)

* Fix #668 (partial)

* Fix requests test

* Read file with retries since cached file may still be being written.

brianbok committed:
WIP

Commit:
Fix analysis completion regression (#696)

* Fix #668 (partial)

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* Fix test regression

* Rename resource

AlexanderSher committed:
Module-scoped imports (#698)

- Make modules and subpackages available for __init__.py packages
- Scope imports to module
- Fixed completion in tensorflow
- Fixed comment
- Fixed GotoDefinition

Commit:
Basic analysis progress reporting (#687)

* Fix #668 (partial)

* Progress reporting

* Tweaked

jakebailey committed:
Fix nullref during testing (#706)

Running analysis tests after #687 fails because there is no progress reporting service.

jakebailey committed:
Compile docstring converter regexes (#707)

* compile docstring converter regexes

* don't bother changing the comments

AlexanderSher committed:
Fix #24: 'Go to Definition' for relative imports not resolving directly to the appropriate module (#702)

* Fix #24: 'Go to Definition' for relative imports not resolving directly to the appropriate module

* Add test

jakebailey committed:
unfilter __all__ (#708)

Fixes #642.
Reopens #619.
Related #620.

This unfilters exported variables except for things which are directly from typing or are imported modules, though even that may be too strong. The old LS didn't look at __all__ whatsoever, so maybe all of these should be allowed to come through; you tell me.

brianbok committed:
Covering more error cases

brianbok committed:
Rename fstring expression and create new tokenkind

brianbok committed:
Expressions public fields changed

brianbok committed:
Using node type instead of expression

Parsing yield as yield statements instead of eating
yield keyworkd

brianbok committed:
order of fields in parseroptions

brianbok committed:
Use of resources

brianbok committed:
Changes to parsing process

brianbok committed:
Parser tests

brianbok committed:
Async walk

brianbok committed:
Using more resource strings

brianbok committed:
Fstring errors

brianbok committed:
Merge branch 'master' into fstrings

brianbok committed:
Parsing errors

brianbok committed:
Tests and handling backslash

brianbok committed:
Error handling

brianbok committed:
Assert

brianbok committed:
Options to parser

brianbok committed:
Resources strings

brianbok committed:
Nit

brianbok committed:
Errors adding with correct line number

brianbok committed:
reusing method in error sink

brianbok committed:
Adding support for setting initial source location in parser (#712)

We make sure that error reports have correct line numbers and columns, considering the initial location.
This will be useful when we create parsers for f-strings sub expressions.

brianbok committed:
Merge branch 'master' into fstrings

AlexanderSher committed:
- Various import completion fixes, import search now is always happening through PathResolver (#722)

- Filtering empty name at the end of module path

jakebailey committed:
Create stub resolver before builtins (#723)

Fixes #719.

Looking at the stack trace on that issue, `Initialize` occurs, which eventually gets an AST, which then spawns a task which eventually calls `CreateModule`, which refers back to `PythonInterpreter`, whose stub resolver has not yet been instantiated. I don't believe instantiating it earlier has any effect other than that it's there earlier so won't be `null`, as `InitializeAsync` only sets up an internal path resolver.

I'm not really sure if this is the "true" fix, since it seems weird for there to be this sort of cycle, but I'm not the most familiar with this specific bit.

brianbok committed:
fixes

brianbok committed:
Merge branch 'master' into fstrings

brianbok committed:
Modeling format specifier

I can now regenerate fstring's code without verbatim mode
save the openQuotes for that

AlexanderSher committed:
Reduce amount of async code in PythonAnalyzer (#718)

* Reduce amount of async code in PythonAnalyzer

* Better English

* - Allow access to AST instance only with corresponding module instance
- Update IPythonModule instance when entry is invalidated

* Add missing `node.Skip()`

* Allow first analysis to be applied even if the analysis version is outdated.

AlexanderSher committed:
Fix #719: Scraper produces incorrect output for the Anaconda 2X builtin (#733)

* Fix #719: Scraper produces incorrect output for the Anaconda 2X builtin

* Add Anaconda 3.7 to the tests interpreters

brianbok committed:
format specifier node

brianbok committed:
Change fstring constructor, use readonly list

brianbok committed:
Array instead of list

brianbok committed:
Tokenizer changes

brianbok committed:
Tokenizer changes

brianbok committed:
Expression location

brianbok committed:
Hovering test

brianbok committed:
Empty lambda issue

Commit:
Fix performance bottleneck fetching module content (#742)

* Fix #668 (partial)

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* Cache document content

Commit:
Optimize scope search and variable lookup (#743)

* Fix #668 (partial)

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* Optimize scope search

* More streamlining

* Optimize typical scope lookup

* PR feedback

brianbok committed:
More tests

brianbok committed:
Lambda parenthesis warning

brianbok committed:
Merge branch 'lambda-parser' into fstrings

brianbok committed:
Error message fix

brianbok committed:
Exact location and better error message

brianbok committed:
Merge branch 'lambda-parser' into fstrings

brianbok committed:
Lambda parenthesis now an error

brianbok committed:
Added test for lambda without parenthesis

brianbok committed:
Format specifier with '!' inside

brianbok committed:
Newline inside expression

brianbok committed:
ParserOptions simplification

brianbok committed:
Unicode escaping tests

brianbok committed:
Another test case

brianbok committed:
Test file was wrong

brianbok committed:
Using 'NoParameters

brianbok committed:
Merge branch 'lambda-parser' into fstrings

brianbok committed:
Tokenizer test

AlexanderSher committed:
Stability improvements (#749)

- Reanalyze user module out of the order if required
- Skip walker if it is outdated and doesn't have a lot of unanalyzed modules.
- Initialize PriorityProducerConsumer<IDependencyChainNode<TValue>> only when GetNextAsync is called for the first time
- Fix ProcessServices so that error stream doesn't hang the process

Commit:
Wire up events from RDT to module resolution (#703)

* Fix #668 (partial)

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* Basic event wire up

* Add test

* Test

* Formatting

* Fix search paths mess

* Handle deleted files better

* PR feedback

* MErge issues

* Fix race condition at builtins load

* PR feedback

Commit:
Update to support IntelliCode (#747)

* Fix #668 (partial)

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* IC update

* usings

* usings

brianbok committed:
Implicitly typed array

brianbok committed:
More implicitly typed arrays

brianbok committed:
Merge branch 'lambda-parser' into fstrings

brianbok committed:
More detail in test

brianbok committed:
Sinplifying changes

brianbok committed:
Sinplifying changes

brianbok committed:
Tests

Commit:
Suppress completion after error tokens (#752)

* Fix #668 (partial)

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* Suppress completion after error tokens

brianbok committed:
Empty lambda before end of file not reporting any error (#741)

We can now report when we don't find colon separator for expression body.
Before, when we had a lambda token followed by end of file, we were not reporting anything.

Commit:
Fix OOB exception in imports (#753)

* Fix #668 (partial)

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* Suppress completion after error tokens

* Fix OOB exception

Commit:
Linter for undefined variables (#693)

* Fix #668 (partial)

* Undef variables, first cut

* Reorg

* Tests

* Tests

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* Options and tests

* Don't squiggle builtin-ins

* Test for function arguments

* Fix tuple assignment in analysis

* Don't false positive on functions and classes forward references

* Disable tracking assignment location
Fix declaration of variables in for loop

* Track multiple locations

* Tests

* using

* Properly look at locations

* Comprehensions etc

* Test update

* Add support for lambdas

* Add lambda linting

* Handle tuple assignment better

* Test update

* Correct comprehension iterator handlint

* Fix race condition at builtins load

* Merge master

* Restore linter

brianbok committed:
Merge branch 'master' into tokenizer-indexes

AlexanderSher committed:
- IPythonAnalyzer.InvalidateAnalysis is now called when first parse happens (#755)

- Fixes for tests that require final analysis while user module (created by RDT.OpenDocument) can have intermediate analysis

brianbok committed:
Errors with initial location

Commit:
Remove old analysis

brianbok committed:
Merge branch 'master' into fstrings

brianbok committed:
Merge branch 'tokenizer-indexes' into fstrings

brianbok committed:
Hovering over incomplete fstrings

brianbok committed:
Test for invalid conversion character

brianbok committed:
Adding types for fstrings

brianbok committed:
Diff looks much smaller now, using datarows

brianbok committed:
Avoiding copy pasting parseErrors method

brianbok committed:
Nit

brianbok committed:
Nit tokenizer tests

brianbok committed:
Merge branch 'tokenizer-indexes' into fstrings

brianbok committed:
FormattedValue is now subtype of node

It used to subtype expression, but it was wrong.

brianbok committed:
Indexes in errors corrected

brianbok committed:
Better error message

brianbok committed:
Merge branch 'tokenizer-indexes' into fstrings

brianbok committed:
Added test case

brianbok committed:
Better error handling`

brianbok committed:
Offset instead of displacement

brianbok committed:
Merge branch 'tokenizer-indexes' into fstrings

brianbok committed:
Parsing string logic looks more like original

brianbok committed:
Tokenizer produces indexes matching absolute index (considering initial location) (#757)

Now TokenSpan and related methods produce indexes starting at the initial location index.
Parsing errors tests that rely on "AllErrors.py" are moved. We check that error location considers initial location for offsetting.
Added test file for Tokenizer.

Commit:
More reliable progress indication (#759)

* Fix #668 (partial)

* Tests

* Revert "Tests"

This reverts commit 7ffc9db4f0807314d4d36a8540380b988c26ee2e.

* Exp

* Exp

* Progress

* Merge fix

* Progress tracker

* Fix merge issue

* Use walker instead

brianbok committed:
More error tests and handling

brianbok committed:
Changing to internal for hiding parser and builder

brianbok committed:
Order fields

brianbok committed:
Renaming method

brianbok committed:
adding comment

brianbok committed:
Merge branch 'master' into fstrings

brianbok committed:
deleted newline

brianbok committed:
Add lambda test case

brianbok committed:
Missing newline

brianbok committed:
Deleting builder: it was overcomplicated

brianbok committed:
Adding tests: no completion inside fstrings

brianbok committed:
Tests: No completion inside fstrings

brianbok committed:
No hovering over fstrings (constant part)

brianbok committed:
Deleting manual index calculating for setting loc

brianbok committed:
Better handling of whitespace

brianbok committed:
Switch instead of dictionary
